fix(ast): estree compat `Function`

hi-ogawa opened the pull request:
Part of https://github.com/oxc-project/oxc/issues/2854

diff of estree_test262.snap

```diff
-Positive Passed: 6233/44293 (14.07%)
+Positive Passed: 13042/44293 (29.44%)
```

hi-ogawa commented:
* **#8972** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8972?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8972?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8972).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hi-ogawa commented on crates/oxc_ast/src/ast/js.rs:
```diff
@@ -1695,9 +1699,11 @@ pub enum FormalParameterKind {
+#[estree(rename = "BlockStatement")]
 pub struct FunctionBody<'a> {
     pub span: Span,
     pub directives: Vec<'a, Directive<'a>>,
+    #[estree(rename = "body")]
     pub statements: Vec<'a, Statement<'a>>,
```
both `directives` and `statements` should be in `body`, but `append_to` macro seems to break for non optional field. I'll look into updating macro for this case separately.

overlookmotel replied:
Yes, extending `#[estree(append_to = ...)]` to cover appending `Vec`s too would be ideal. We could alternatively add `#[estree(prepend_to = ...)]`, which would make more sense here. But I think for now we already have enough attributes - probably better to use `append_to` at least for now.

Let me know if you have trouble working out how to adapt the codegen. I think it *mostly* makes sense, but I wrote it, so of course it makes sense *to me*!

overlookmotel reviewed (approved):
Great!

Side note: I'm wondering in what circumstances the `expression` field is `true`?

Seems not to be `true` for function expressions ([AST explorer](https://astexplorer.net/#/gist/34184c71bf0517d2df35cd4164fd2dd6/6e835451ada5c1172cb1a0aab32ba60f672335fe)). So when *is* it true? I guess we'll find out as we progress through the test cases...

overlookmotel commented:
### Merge activity

* **Feb 9, 1:52 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

hi-ogawa committed:
fix(ast): estree compat `Function` (#8972)

Part of https://github.com/oxc-project/oxc/issues/2854

diff of estree_test262.snap

```diff
-Positive Passed: 6233/44293 (14.07%)
+Positive Passed: 13042/44293 (29.44%)
```

hi-ogawa commented:
> Side note: I'm wondering in what circumstances the `expression` field is `true`?

I just looked into estree and this is probably for some ancient deprecated feature https://github.com/estree/estree/blob/2b48e56efc223ea477a45b5e034039934c5791fa/deprecated.md#functions

In https://bugzilla.mozilla.org/show_bug.cgi?id=1083458, I found a weird syntax like `function(l) null` for the first time :open_mouth:

overlookmotel commented:
> I just looked into estree and this is probably for some ancient deprecated feature https://github.com/estree/estree/blob/2b48e56efc223ea477a45b5e034039934c5791fa/deprecated.md#functions
> 
> In https://bugzilla.mozilla.org/show_bug.cgi?id=1083458, I found a weird syntax like `function(l) null` for the first time 😮

Ah ha! `function(l) null` - thank god this did not take off!
